src/xserver.cpp

```diff
@@ -0,0 +1,71 @@
+/***************************************************************************
+* Copyright (c) 2019, Martin Renou, Johan Mabille, Sylvain Corlay and      *
+* Wolf Vollprecht                                                          *
+*                                                                          *
+* Distributed under the terms of the BSD 3-Clause License.                 *
+*                                                                          *
+* The full license is in the file LICENSE, distributed with this software. *
+****************************************************************************/
+
+#include "xeus/xserver_zmq.hpp"
+
+#include "xserver.hpp"
+
+#include "pybind11/pybind11.h"
+
+namespace py = pybind11;
+
+namespace xpyt
+{
+    server::server(zmq::context_t& context, const xeus::xconfiguration& config)
+        : xeus::xserver_zmq(context, config)
+    {
+    }
+
+    server::~server()
+    {
+    }
+
+    void server::start_impl(zmq::multipart_t& message)
+    {
+        start_publisher_thread();
+        start_heartbeat_thread();
+
+        m_request_stop = false;
+
+        publish(message);
+
+        while (!m_request_stop)
+        {
+            poll_msg();
+        }
+
+        stop_channels();
+
+        std::exit(0);
+    }
+
+    void server::poll_msg()
+    {
+        zmq::pollitem_t items[]
+            = { { m_controller, 0, ZMQ_POLLIN, 0 }, { m_shell, 0, ZMQ_POLLIN, 0 } };
+
+        py::gil_scoped_release release;
+        zmq::poll(&items[0], 2, -1);
+        py::gil_scoped_acquire acquire;
```

Comment:
I am now releasing the GIL when waiting for new messages, allowing Python secondary threads to resume their job

Reply:
`gil_scoped_release` and `gil_scoped_acquire` are scoped guard, meaning they are meant to be used in a scope, their destructor will do the opposite operation of the one in their constructor when the thread exists the scope. So the code should be
```cpp
{
    py::gil_scoped_release release;
    zmq::poll(&items[0], 2, -1);
}
```

Reply:
@martinRenou why don't you remove all of the xserver here? You can release the GIL after you finish the execute_request and acquire the GIL when you get the next execute request. This should be only a 2 line change, no?

Reply:
Yes I will. Did not have time to do it yet.

Reply:
I need to put a `py::gil_scoped_acquire` for every request where I execute Python code. But I also need to check if the acquire works like a reentrant mutex, if it's the case we need to be careful that the GIL was not acquired earlier.

Reply:
Mmmmh... It is not as simple as I thought it would be. The acquire crashes.
```
Program received signal SIGSEGV, Segmentation fault.
x00007ffff75cab80 in threadstate_getframe ()

traceback:
#0  0x00007ffff75cab80 in threadstate_getframe () from /home/martin/miniconda3/envs/xeus-python/lib/libpython3.7m.so.1.0
#1  0x00007ffff7658caa in PyEval_GetBuiltins () from /home/martin/miniconda3/envs/xeus-python/lib/libpython3.7m.so.1.0
#2  0x00007ffff77afa25 in pybind11::detail::get_internals() () from /home/martin/miniconda3/envs/xeus-python/lib/libxeus-python.so.0
#3  0x00007ffff77b0539 in pybind11::gil_scoped_acquire::gil_scoped_acquire() [clone .lto_priv.253] () from /home/martin/miniconda3/envs/xeus-python/lib/libxeus-python.so.0
```

Reply:
Ok it's fixed now